common/simple_dom.h

```diff
@@ -0,0 +1,274 @@
+/*
+Copyright 2022 The Photon Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#pragma once
+#include <inttypes.h>
+#include <assert.h>
+#include <memory>
+#include <math.h>
+#include <atomic>
+#include <photon/common/object.h>
+#include <photon/common/estring.h>
+#include <photon/common/enumerable.h>
+#include <photon/common/stream.h>
+
+
+namespace photon {
+
+namespace fs {
+class IFileSystem;
+}
+
+// simple unified interface for common needs in reading
+// structured documents such as xml, json, yaml, ini, etc.
+namespace SimpleDOM {
+
+// SimpleDOM emphasize on:
+// 1. simple & convenient usage;
+// 2. unified interface, common needs (may not fit for corner cases);
+// 3. efficient parsing (reading), as well as compiling;
+
+using str = estring_view;
+
+class RootNode;
+struct NodeWrapper;
+
+// the interface for internal implementations
+class Node : public Object {
+protected:
+    Node() = default;
+    Node* _parent;
+union {
+    Node* _next;
+    const char* _text_begin;     // the root node have text begin (base
+};  rstring_view32 _key, _value; // of _key and _value of rstring_view)
+
+    friend struct NodeWrapper;
+
+public:
+    virtual size_t num_children() const __attribute__((pure)) = 0;
+
+    // get the i-th child node
+    virtual Node* get(size_t i) const __attribute__((pure)) = 0;
+
+    // get the first child node with a specified `key`
+    virtual Node* get(str key) const __attribute__((pure)) = 0;
+
+    RootNode* root() const __attribute__((pure)) {
+        auto ptr = this;
+        while (ptr->_parent)
+            ptr = ptr->_parent;
+        return (RootNode*)ptr;
+    }
+
+    NodeWrapper wrapper();
+};
+
+class RootNode : public Node {
+protected:
+    RootNode() = default;
+
+    // reference counting is only applied to root
+    // node, which represents the whole document
+    std::atomic<uint32_t> _refcnt{0};
+
+public:
+    void add_ref() { ++_refcnt; }
+    void del_ref()  { if (--_refcnt == 0) delete this; }
+};
+
+#define IF_RET(e) if (_node) return e; \
+                        else return {};
+
+// the interface for users
+struct NodeWrapper {
```

Comment:
For user use, it's better to be named as "Node". The original "Node" can be renamed to NodeBase or something else, and not be exposed externally.

Reply:
OK